### PR TITLE
Add addCredentials operation

### PR DESCRIPTION
**Pull request by marians**
Towards
- https://github.com/giantswarm/giantswarm/issues/2944
- https://github.com/giantswarm/giantswarm/issues/2652 (epic)

This PR adds the spec for a `addCredentials` operation that allows to set the provider credentials for an organization.

### Preview

For an AWS installation and an organization named `acme`, this is how credentials could be specified:

```nohighlight
curl -X POST -H "Authorization: giantswarm $TOKEN" $API_ENDPOINT/v4/organizations/acme/credentials/ -d '
{
  "provider": "aws",
  "aws": {
    "roles": {
      "admin": "arn:aws:iam::123456789012:role/GiantSwarmAdmin",
      "awsoperator": "arn:aws:iam::123456789012:role/GiantSwarmAWSOperator"
    }
  }
}'
```

Response:

```nohighlight
HTTP/1.1 201 Created
Location: /v4/organizations/acme/credentials/abc123/
Content-type: application/json

{
  "code": "RESOURCE_CREATED",
  "message": "A new set of credentials has been created with ID 'abc123'"
}
```

### CODE REVIEW

File: definitions.yaml [hunk 1]
```diff
@@ -568,3 +568,31 @@ definitions:
...
+      - provider
+    description: Request model for adding a set of credentials
+    properties:
+      provider:
+        type: string
+      description:
+        type: string
+      aws:
+        type: object
+        description: Credentials specific to an AWS
```
Comment (TheoBrigitte) [inline definitions.yaml]: to an AWS ... ? account?

File: definitions.yaml [hunk 2]
```diff
@@ -568,3 +568,31 @@ definitions:
...
         description: The date and time when this account will expire
+
+  V4AddCredentialsRequest:
+    type: object
+    required:
+      - provider
+    description: Request model for adding a set of credentials
+    properties:
+      provider:
+        type: string
```
Comment (xh3b4sd) [inline definitions.yaml]: Why do we need the provider when we have the `aws` block?

File: spec.yaml [hunk 3]
```diff
@@ -997,6 +997,58 @@ paths:
...
             $ref: "./definitions.yaml#/definitions/V4GenericResponse"
 
+  /v4/organizations/{organization_id}/credentials/:
+    post:
+      operationId: addCredentials
+      tags:
+        - organizations
+      summary: Set credentials
+      description: |
+        Add a set of credentials to the organization allowing to create and
```
Comment (rossf7) [inline spec.yaml]: "allowing to" sounds a bit strange to me. "allowing the creation and operation of clusters" is better IMO.

**Comment by oponder** [comment]
> "Credentials in an organization are immutable. An organization can only have one set of credentials."


Have we addressed how we'll edit credentials then if the user makes a typo?

**Comment by marians** [comment]
> Have we addressed how we'll edit credentials then if the user makes a typo?

This is beyond the API in the first iteration. The workaround will require creating a new organization.

**Comment by JosephSalisbury** [comment]
@marians do you have some thoughts on error responses?

- if the organization does not exist?
- if the provider key != aws?
- if the aws key is empty?
- if the roles list is empty?
- if the roles list does not have two entries?
- if the arn key is not valid?

**Comment by kopiczko** [comment]
Also Secret format will be close to that: https://github.com/giantswarm/giantswarm/issues/3070#issuecomment-385345083

Reason: simplified validation. Something to think about here.

File: spec.yaml [hunk 4]
```diff
@@ -997,6 +997,78 @@ paths:
...
+        ```json
+        {
+          "provider": "aws",
+          "aws": {
+            "roles": {
+              "admin": "arn:aws:iam::123456789012:role/GiantSwarmAdmin",
+              "awsoperator": "arn:aws:iam::123456789012:role/GiantSwarmAWSOperator"
+            }
+          }
+        }
```
Comment (kopiczko) [inline spec.yaml]: <3

**Comment by marians** [comment]
Merging as this is implemented. Thanks folks!

**Pull request merged**